src/OrchardCore.Modules/OrchardCore.Media/Services/MediaFileStoreResolverMiddleware.cs

```diff
@@ -0,0 +1,180 @@
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace OrchardCore.Media.Services
+{
+    /// <summary>
+    /// Provides local caching for remote file store files.
+    /// </summary>
+    public class MediaFileStoreResolverMiddleware
+    {
+        // Use default stream copy buffer size to stay in gen0 garbage collection;
+        private const int StreamCopyBufferSize = 81920;
+
+        private static readonly ConcurrentDictionary<string, Lazy<Task>> _writeTasks = new ConcurrentDictionary<string, Lazy<Task>>();
```

Comment:
This was the best way I could come up with to handle multiple requests for the same file.
Thought it might be better than an asynckeylock arrangement, as it only gets used (other than a dictionary lookup that will mostly no-op) when actually saving a file from blob to file cache.

Reply:
This is much better than an async lock. We do this in the Cache TagHelper to prevent multiple queries from building the same cached entry, and I also do it in YesSql to reuse concurrent queries that are "in flight". Well done!

Reply:
Great, thank you. Feels like I'm the right track here.